### PR TITLE
Add tests for nil sender/receiver

### PR DESCRIPTION
**Pull request by pkwarren**
In the HTTP handler, there is code that handles returning nil
sender/receiver and continues to call interceptors. This is currently
unreachable with the exported API - add tests using a custom
protocolHandler implementation.

### CODE REVIEW

File: handler.go [hunk 1]
```diff
@@ -274,6 +276,10 @@ func (h *Handler) ServeHTTP(responseWriter http.ResponseWriter, request *http.Re
 		receiver = interceptor.WrapStreamReceiver(ctx, receiver)
 	}
 	h.implementation(ctx, sender, receiver, clientVisibleError)
+	if sendHTTPInternalError {
+		// protocolHandler returned a nil sender, so we can't send a proper HTTP response.
+		responseWriter.WriteHeader(http.StatusInternalServerError)
```
Comment (pkwarren) [inline handler.go:281]: If the sender is nil, that means we wrap a no-op sender (which sends nothing across the wire). In that case, we should at least send back an error from the handler. Without this logic, we'd send back an empty HTTP 200 response.

I'm not sure if we want to encode more details about the `clientVisibleError` in a response body here but this should at least reflect that an error occurred on the client.

File: handler_test.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	pingv1 "github.com/bufbuild/connect-go/internal/gen/connect/ping/v1"
+)
+
+var errFailedToCreateStream = errors.New("failed to create stream")
+
+func TestInterceptorsCalledIfSenderReceiverNil(t *testing.T) {
+	t.Parallel()
+	mux := http.NewServeMux()
+	pingPath := "/connect.ping.v1.PingService/Ping"
+	handler := NewUnaryHandler[pingv1.PingRequest, pingv1.PingResponse](
```
Comment (pkwarren) [inline handler_test.go:34]: Have to use NewUnaryHandler here directly since the test needs to access the unexported `protocolHandlers` field so we can't use the connect generated code (import cycle).

File: handler_test.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		pingPath,
+		func(ctx context.Context, request *Request[pingv1.PingRequest]) (*Response[pingv1.PingResponse], error) {
+			t.Error("shouldn't call handler implementation")
```
Comment (pkwarren) [inline handler_test.go:37]: We shouldn't call the unary method if we get an error creating the stream - from reading the code, the expectation is that we should only call interceptors.

File: handler_test.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+			return nil, NewError(CodeUnimplemented, nil)
+		},
+		WithInterceptors(UnaryInterceptorFunc(func(next UnaryFunc) UnaryFunc {
+			return func(ctx context.Context, request AnyRequest) (AnyResponse, error) {
+				_, err := next(ctx, request)
+				assert.NotNil(t, err)
+				assert.True(t, errors.Is(err, errFailedToCreateStream))
+				return nil, err
```
Comment (pkwarren) [inline handler_test.go:45]: Validate that the interceptor is called with the clientVisibleError.

Note that this might limit what an interceptor can really do since changing the response or returning a custom error won't have any impact at the client (if sender is nil). The request will also be an empty request and not very useful if the receiver is nil. This will only be useful for a limited set of interceptors (i.e. logging/tracing).

File: handler_test.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+			}
+		})),
+	)
+	// Override gRPC handler with custom one
+	handler.protocolHandlers = append([]protocolHandler{&nilProtocolHandler{}}, handler.protocolHandlers...)
```
Comment (pkwarren) [inline handler_test.go:50]: For now, have to manually access unexported field to trigger the code path in the handler when the sender/receiver are nil.

**Review by pkwarren — COMMENTED** [review]
This is an alternative to https://github.com/bufbuild/connect-go/pull/290 and should verify the code path if we ever introduce a new protocol handler in the future that returns a nil sender/receiver or open the API to add new protocol handlers.

File: handler.go (hunk 1 above)
Comment (amckinney) [reply handler.go:281]: It looks like the `clientVisibleError` could be a typed `*connect.Error`. If so, we should map its error code to the associated HTTP code rather than always using `http.StatusInternalError`. For example, I noticed that `clientVisibleError` is sometimes set to to the `timeoutErr` (which will often be initialized with `CodeInvalidArgument`).

However, the `clientVisibleError` is a detail of the `protocolHandler` that was used to serve the request (e.g. gRPC, Connect, etc). Fortunately, the gRPC and Connect [error codes](https://connect.build/docs/protocol/#error-codes) match, so we _should_ be able to use the same code transformation regardless of the `protocolHandler` that was used. I'd need to take a closer look at this though to refresh.

```go
responseWriter.WriteHeader(connectCodeToHTTP(CodeOf(clientVisibleError)))
```

However, this might be another place for concern if we ever expose another `protocolHandler` that doesn't have the same `connect.Code` -> HTTP Code transformation. We'll need to verify this behavior before we move forward here - I have a feeling we might not want to use the `responseWriter.WriteHeader` call in this context.

**Comment by akshayjshah** [comment]
Mea culpa - I've steered you wrong here. :sob: #290 is definitely the better approach.
